Script for maintainers

basakerdamar opened the pull request:
This script allows to parse MAINTAINERS file using LinuxMaintainers class, reports subsystems or maintainers optionally filtered by a given list of files together with LoC and byte counts for the files.
A command like: pasta maintainers show [--smallstat] [--outfile OUTFILE] [--file FILE] [--filter FILTER] would provide the option to filter by the contents of FILTER, allow to use another MAINTAINERS file FILE other than the one on the repository, and see a smaller view with less fields in the report.

rralf commented:
I'm missing a description and the context -- what are you trying to achieve? Thanks!

bulwahn commented:
@rralf, let Anmol and me first go through a code review with @basakerdamar; when we are happy with the code, you can do a second pass and give some more hints on things we might have overlooked.

rralf commented:
Alright!

basakerdamar committed:
Integration into pasta

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,165 @@
+'''
+This work is licensed under the terms of the GNU GPL, version 2.  See
+the COPYING file in the top-level directory.
+'''
```
Add a copyright header and author as in https://github.com/lfd/PaStA/blob/master/bin/pasta_evaluate_patches.py

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,165 @@
+    return [item for sublist in nested for item in sublist]
+
+def shorten(string_list):
+    lenght = len(string_list)
```
typo it is length, not lenght.

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,165 @@
+def flatten(nested):
+    return [item for sublist in nested for item in sublist]
+
+def shorten(string_list):
```
Function is not used anymore.

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,165 @@
+    except:
+        return [0, 0]
+
+all_filenames = [os.path.join(r, item)[23:] for r, _, f in os.walk('./resources/linux/repo') for item in f]
```
You should not hard-code the subpath, you should take that from the pasta config. Of course, it only makes sense if the config is a linux config anyway.

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,165 @@
+        subsystems = all_maintainers.get_subsystems_by_file(filename)
+        for subsystem in subsystems:
+            subsystem_dirs[subsystem] = subsystem_dirs.get(subsystem, set())
+            subsystem_dirs[subsystem].add(''.join(re.findall('[a-zA-Z0-9_]+/', filename)))
```
what is this regular expression for?

basakerdamar replied:
This was to get directories, I will use a built-in function instead

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,165 @@
+from logging import getLogger
+from collections import OrderedDict
+from argparse import ArgumentParser
+import numpy as np
```
shortening it to np is really overdoing it, I think.

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,165 @@
+        for subsystem in subsystems:
+            maintainers = all_maintainers.get_maintainers(subsystem)
+            for maintainer in flatten(maintainers):
+                if not (type(maintainer) == str or maintainer[0] == '' or maintainer[1] == ''):
```
If that does not fit, you should at least about it in some verbose mode, right?

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,165 @@
+            maintainers = all_maintainers.get_maintainers(subsystem)
+            for maintainer in flatten(maintainers):
+                if not (type(maintainer) == str or maintainer[0] == '' or maintainer[1] == ''):
+                    maintainers_subs[maintainer[0]+' <'+maintainer[1]+'>'] = maintainers_subs.get(maintainer[0]+' <'+maintainer[1]+'>', set())
```
maintainer[0]+' <'+maintainer[1]+'>' is a pretty-printing function or a to_string, or string_of_maintainer function.

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,165 @@
+    if args.file!= None:
+        all_maintainers_text = file_to_string(args.file)
+    else:
+        all_maintainers_text = file_to_string('./resources/linux/repo/MAINTAINERS')
```
is there something like a default function in python default(args.file, b) means args.file if that is not None else b.

bulwahn replied:
Then this can be expressed less verbose.

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,165 @@
+        if args.filter != None:
+            counts_filt, _ = get_maintainer_counts(all_maintainers, filenames)
+    else:
+        return -1
```
Probably here you should warn about wrong use and indicate how to get some help.

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,165 @@
+            results.append([title, 'Status', 'LoC', 'Byte count'])
+            for item in counts:
+                if any(all_maintainers.subsystems[item].status):
+                    results.append([item, all_maintainers.subsystems[item].status[0].value, counts[item][0], counts[item][1]]) 
```
all_maintainers.subsystems[item].status[0].value, why is this status[0], could there actually be multiple status entries?

basakerdamar replied:
There seems to be always only one but returned as a list from LinuxSubsystem

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,165 @@
+            results.append([title, 'LoC', 'Byte count', 'Entries'])
+            for item in counts:
+                results.append([item, counts[item][0], counts[item][1], ', '.join(maintainers_subs[item])])
+        
```
you can use a dictionary with lambda functions, and then rewrite this code, so that the four cases are actually just the assignment of four different lists.

bulwahn replied:
and a generic part that handles the rest.

bulwahn reviewed (changes requested):
Here are some comments for improvements.

rralf commented:
@ditzlike this might be interesting for you.

bulwahn commented:
- how about preparing some tests to check if this command works as expected?

- for presentation of entries in show maintainers, sort entries by size themselves

and add a percentage of size of the entry compared to the overall responsibility of that maintainer.
like SCHEDULER (80%), FUTEX (15%), XYZ DRIVER (5%)

When multiple entries are there, and especially when they are cut off, this would be helpful:

(\<number of entries>) entries ...

if number of entries = 1, just drop that information.

- I did not understand "Other directories"

How about:

  - RELEVANT: list of the most general file patterns such that all files with that pattern are included in the relevant list and are in the entry.

  - IRRELEVANT: list of the most general file pattern such that all files under that directory are excluded from the relevant list, but are in the entry.

Then, I think that should cover all cases...

bulwahn commented:
@rralf Do you understand what we are collecting with Basak's script? Otherwise, let us quickly talk.

rralf commented:
> @rralf Do you understand what we are collecting with Basak's script? Otherwise, let us quickly talk.

Hmm, I'm not sure - do you try to imitate get_maintainers.pl? BTW, @ditzlike currently writes some tests that compares output of get_maintainers with PaStA's interpretation of the MAINTAINERS file.

Is this series already ready for my review? I guess I'll understand what you're trying to achieve if I look at the code.

bulwahn commented:
No, we are building some basic scripts to aggregate who is "most important" based on the LoC the maintainer takes care of, and I have another script that provides the list of files for a specific kernel build, so we can create this list of most important maintainers for a kernel build.

Next step, I have in mind, is to also visualize the maintainer hierarchy based on the entries and which entry is a superset of another. Again, maybe Wolfgang already such things in the past?

bulwahn commented:
Ralf, you can always have a look... I will do another round of review with Basak, but we are getting closer to nice clean code in the script.

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -72,9 +60,11 @@ def get_maintainer_counts(all_maintainers, filenames = all_filenames):
+                    maintainers_subs[pretty_name(maintainer)].add(subsystem)
+                    counts_by_maintainer[pretty_name(maintainer)] = add(counts_by_maintainer.get(pretty_name(maintainer), [0, 0]), file_lengths[filename])
+                elif verbose:
+                    print('Ignoring mailing list:', maintainer)
```
I think pasta has a logging interface to use.

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -86,72 +76,68 @@ def get_small(title, counts):
+        cases['entries', False] =[lambda item : [item, counts[item][0], counts[item][1], status(item)], [args.query.capitalize(), 'LoC', 'Byte count','Status']]
+        cases['maintainers', True] =[lambda item : [item, counts_filt[item][0], counts[item][0], counts_filt[item][1], percentage(item), ', '.join(maintainers_subs[item])], [args.query.capitalize(), 'LoC in list', 'Total LoC', 'Byte count',  'In list(%)', 'Entries']]
+        cases['maintainers', False] =[lambda item : [item, counts[item][0], counts[item][1]], [args.query.capitalize(), 'LoC', 'Byte count']]
+
```
Now, look what are duplicate lambda functions and what is the configurable part here? If you split it accordingly, you got a much nicer solution.

bulwahn replied:
I actually meant a dictionary for 'field_name' => function, so 'LoC' => lambda item: counts[item][0]... etc.

basakerdamar replied:
>     * I did not understand "Other directories"
> 
> 
> How about:
> 
>     * RELEVANT: list of the most general file patterns such that all files with that pattern are included in the relevant list and are in the entry.
> 
>     * IRRELEVANT: list of the most general file pattern such that all files under that directory are excluded from the relevant list, but are in the entry.
> 
> 
> Then, I think that should cover all cases...

Hello, 
Other directories are showing the directories that are excluded in the filter list for each of the entries. Would it be meaningful to display some aggregation of these directories in this case?

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -86,72 +76,68 @@ def get_small(title, counts):
+        cases['maintainers', True] =[lambda item : [item, counts_filt[item][0], counts[item][0], counts_filt[item][1], percentage(item), ', '.join(maintainers_subs[item])], [args.query.capitalize(), 'LoC in list', 'Total LoC', 'Byte count',  'In list(%)', 'Entries']]
+        cases['maintainers', False] =[lambda item : [item, counts[item][0], counts[item][1]], [args.query.capitalize(), 'LoC', 'Byte count']]
+
+        if 'counts_filt' not in locals():
```
I think that is a bit too much of meta-programming here. Consider just having one variable above.

bulwahn reviewed (changes requested):
We are getting there...

rralf commented:
Let's Cc Wolfgang. @wolfgangmauerer 

I understand. Two things before my first round of review:
  - If we only need 'add' from numpy, then I'd like to substitute it with something else. Numpy is a beast of a dependency that we probably can avoid, as long as there's no strong reason for it.
  - Basak, could you please rebase on next, and reorder your commits? E.g., first add the stubs to pasta, then add the actual business logic. Makes it easier for me to review

Thanks!
  Ralf

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,174 @@
+"""
+PaStA - Patch Stack Analysis
+Copyright (c) Bayerische Motoren Werke Aktiengesellschaft (BMW AG), 2019
```
It is 2020 already, either change it to 2020 or just drop it :)

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,174 @@
+    fields['Other directories'] = lambda item : ', '.join(dirs[item])
+    fields['Entries of maintainer'] = lambda item : ', '.join(maintainers_subs[item])
+       
+    if args.smallstat:
```
If you look how these lists are composed, you will find that you can compose this list one by one nicely.
E.g.,
title = [a, b if condition, c, d if condition, e if condition, ...]

bulwahn reviewed (commented):
Some more feedback

basakerdamar commented:
> Could you please provide an example how I can exemplarily run this script?

Hello, 

 ./pasta show_maintainers --entries

should show you a view with subsystems and respective sums of file sizes and

./pasta show_maintainers

brings a similar view, this time with maintainers.

You could give a file containing a list with --filter so that the ratios in-list/in-repo are also displayed. In the filter file, file paths should be relative to the repository like arch/x86/boot/a20.c.

I changed the commands a bit for the sake of simplicity. I hope the script also looks a bit better with the improvements.

rralf commented:
Great, thank you very much! Will try that tomorrow.

rralf commented:
> > Could you please provide an example how I can exemplarily run this script?
> 
> Hello,
> 
> ./pasta show_maintainers --entries
> 
> should show you a view with subsystems and respective sums of file sizes and
> 
> ./pasta show_maintainers
> 
> brings a similar view, this time with maintainers.
> 
> You could give a file containing a list with --filter so that the ratios in-list/in-repo are also displayed. In the filter file, file paths should be relative to the repository like arch/x86/boot/a20.c.
> 
> I changed the commands a bit for the sake of simplicity. I hope the script also looks a bit better with the improvements.

Basak, thanks for the pointer, I think begin to understand what you try to achieve.

bulwahn commented:
Basak, the feature listing 

RELEVANT: list of the most general file patterns such that all files with that pattern are included in the relevant list and are in the entry.

IRRELEVANT: list of the most general file pattern such that all files under that directory are excluded from the relevant list, but are in the entry.

is still missing, right?

If it is not clear at the moment, let us just the dump the whole list of files included and excluded and go through a few examples to see how that can be nicely aggregated.

basakerdamar commented:
For both RELEVANT and IRRELEVANT, I was implementing most common containing folder(s) but it did not seem like a smart solution as an aggregation. I will bring back the list of directories for now.

bulwahn commented:
Okay, I will have a look.

bulwahn commented:
Can we include the list of files that are considered relevant and irrelevant?

I think there is some kind of bug in showing the directories; I simply could not make sense of why it was either in relevant or irrelevant when it was actually mixed when cross-checking the list of relevant files and the description in MAINTAINERS.

bulwahn commented:
I think we can also drop the byte count by default, it confuses me more than it helps.

rralf reviewed (commented):
The files that are affected by a specific configuration should go to the resources/ directory. @bulwahn It'd also be great if we could include the script that creates the file.

Ah, and we could simply us the file extension .files, and think of a standardises naming scheme. We can use information of the filename to tell PaStA which version of MAINTAINERS it should load. Basak, your filename already contains "v5.6", which is very useful.

We could (just a suggestion) use a dot-separation like:

config-whatever-x86-machine.v5.5-rc3.files

For reasons traceability, I'd also like to store the config itself.

rralf commented:
For the rest of the PR: Could you please make a clean patch stack out of it? Makes it easier for me to review. It'd be great if there would be just two commits: The first one that introduces the stubs for the new tool, and a second one that adds the business logic on top. The second one should introduce a working state of your tool. Any further changes/improvements can be made on top of that.

rralf commented on pasta:
```diff
@@ -43,7 +43,7 @@ from bin.pasta_patch_descriptions import patch_descriptions
 from bin.pasta_upstream_history import pasta_upstream_history
 from bin.pasta_web import web
-from bin.pasta_show_maintainers import get_maintainers
+from bin.pasta_show_maintainers import show_maintainers
```
No need to change this line, if you use the final name (show_maintainers) in the first commit.

rralf commented on pasta:
```diff
@@ -171,8 +171,8 @@ def main(argv):
-    elif sub == 'maintainers':
-        return get_maintainers(config, sub, argv)
+    elif sub == 'show_maintainers':
+        return show_maintainers(config, sub, argv)
```
same here

rralf reviewed (commented):
Please use the final name show_maintainers in this commit

rralf commented:
Thanks! Will have a deeper look into the second commit tomorrow!

bulwahn commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,286 @@
+            title += ['LoC in list', 'In list(%)']
+            if args.entries:
+                title += ['Irrelevant directories', 'Relevant directories']
+        if args.bytes:
```
If you move this a bit above, you can simply use append or title += ['Byte count'].

bulwahn commented:
> The files that are affected by a specific configuration should go to the resources/ directory. @bulwahn It'd also be great if we could include the script that creates the file.
>

On my todo list for weeks now, it is a simple script, but I need to document it properly.
 
> Ah, and we could simply us the file extension .files, and think of a standardises naming scheme. We can use information of the filename to tell PaStA which version of MAINTAINERS it should load. Basak, your filename already contains "v5.6", which is very useful.
> 
> We could (just a suggestion) use a dot-separation like:
> 
> config-whatever-x86-machine.v5.5-rc3.files
> 
> For reasons traceability, I'd also like to store the config itself.

We can surely do that.

bulwahn commented:
> Any further changes/improvements can be made on top of that.

I would still like to see the relevant and irrelevant files/directories as a feature in a way it is useful for me. We can commit it to the repository in the current state, but I still think it needs some improvements (or different options to be able to query what the user really wants to know) afterwards.

rralf reviewed (changes requested):
You just deleted the resources submodule. What i meant was to move the config to the resources submodule.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,299 @@
+                                          '(default: %(default)s)')
+    args = parser.parse_args()
+    d_repo = config.repo_location
+    repo = pygit2.Repository(d_repo)
```
repo = config.repo.repo (we already have a Repository instance)

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,299 @@
+        all_maintainers = file_to_string(args.file)
+        LinuxMaintainers(all_maintainers)
+    else:
+        if kernel_version:
```
if !args.filter -> referenced before assignment

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,299 @@
+    repo = pygit2.Repository(d_repo)
+    commit_version = args.commit if args.commit else 'HEAD'
+    commit = repo.revparse_single(commit_version)
+    tree = commit.tree
```
Hmm... Fails if I use --commit v5.3. Would be nice if we could support both.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,299 @@
+    parser.add_argument('--file', type=str, help='Specify MAINTAINERS file')
+    parser.add_argument('--filter', type=str, help='Filter by file list')
+    parser.add_argument('--commit', type=str, 
+                            help='Specify a commit hash for Linux repo')
```
Change --commit to --revision. --revision should at least support a specific commit hash and a tree object. And we should properly handle errors.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,299 @@
+    parser.add_argument('--smallstat', action='store_true', help='Simple view')
+    parser.add_argument('--bytes', action='store_true', help='Show byte counts')
+    parser.add_argument('--outfile', type=str, help='Output to a csv file')
+    parser.add_argument('--file', type=str, help='Specify MAINTAINERS file')
```
Do we still need that argument? I guess we can drop it. It's totally sufficient if we can specify the kernel version.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,299 @@
+    args = parser.parse_args()
+    d_repo = config.repo_location
+    repo = pygit2.Repository(d_repo)
+    commit_version = args.commit if args.commit else 'HEAD'
```
If we encode the version in the filename, then we should use this information as the default value instead of HEAD. --revision can be used to override the version.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,299 @@
+    global _tmp_tree
+    _tmp_tree = tree
+
+    processes = int(cpu_count() * args.cpu_factor)
```
Did you experience any scaling issues? If not, then I'd say we can drop the cpu_factor.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,299 @@
+                print(line_s[0:columns-3]+"...")
+            else:
+                print(line_s)
+    return 0
```
Nit: Missing newline at EOF. And please keep an eye on trailing spaces. You can, for example, use git show to detect them.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,299 @@
+
+    if args.file:
+        all_maintainers = file_to_string(args.file)
+        LinuxMaintainers(all_maintainers)
```
... Looks incomplete. However, I guess we can drop it entirely.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,299 @@
+    parser.add_argument('--bytes', action='store_true', help='Show byte counts')
+    parser.add_argument('--outfile', type=str, help='Output to a csv file')
+    parser.add_argument('--file', type=str, help='Specify MAINTAINERS file')
+    parser.add_argument('--filter', type=str, help='Filter by file list')
```
Please try to improve the description of this argument. I had to walk through the code in order to understand, that --filter needs to point to the defconfig.files file.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,299 @@
+    if args.filter:
+        filter_by_files = file_to_string(args.filter).splitlines()
+        # Automatically find kernel version from file name "v5.6"
+        version_regex = re.compile(r'v[0-9]+.[0-9]+')
```
Simply split the filename at the dots. Then, there's no need for a regex (that does, for example, currently not work for v5.4-rc3).

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,299 @@
+        else:
+            all_maintainers = tree['MAINTAINERS'].data.decode('iso8859')
+
+    all_filenames = walk_commit_tree(tree)
```
Ok, just a second. Let's get one step back and think of our initial problem:

We want to know the scale/volume of subsystems of the kernel. Either from an overall perspective, or related to a specific configuration. A specific configuration is always mapped to a version and list of affected files. So it is not necessary to specify both: Kernel version and a commit hash for the commit tree walk.

Let me give you two examples:

Get an overall overview of the maintainers of v5.3
$ ./pasta show_maintainers --entries --revision v5.3

Get an overall overview of the maintainers of HEAD
$ ./pasta show_maintainers --entries

Get an overview of the maintainers for a specific configuration, will default to MAINTAINERS of v5.4
$ ./pasta show_maintainers --entries --config defconfig-x86.v5.4.files

Get an overview of the maintainers for a specific configuration, will enforce v5.5, though (probably never required)
$ ./pasta show_maintainers --entries --config defconfig-x86.v5.4.files --revision v5.5

rralf reviewed (changes requested):
Hi Basak,

thank you very much for your work! I can see the improvements in every round. So please don't be confused that there are still so many comments. That's not unusual. In this round, I only focused on the configuration. I did not dive into other parts of your script. I will do that once we have a common understanding of how to run/configure the script.

Thanks
  Ralf

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -193,42 +204,35 @@ def show_maintainers(config, sub, argv):
         # Automatically find kernel version from file name "v5.6"
-        version_regex = re.compile(r'v[0-9]+.[0-9]+')
-        kernel_version = version_regex.findall(args.filter)[0]
+        extract_version = '.'.join(args.filter.split('.')[1:3])
```
Yikes, I just realised that dot is probably not the best choice as separator :-)

rralf replied:
Let me try another approach, will get back to you.

basakerdamar replied:
That was just my work in progress, sorry about the mess :) 
Should I go back to regex?

rralf replied:
No worries. I'm not sure let. I'm playing around with some possibilities. Will provide you a patch for on top of your series.

basakerdamar committed:
Maintainers script & Relevant and irrelevant file lists for subsystems view with filtering

basakerdamar commented:
I think that I used the wrong command to move the relevant-files which ended up deleting the resources submodule. Sorry about the confusion! I will fix that now.

rralf commented:
Hold on, I've already a local workaround. Will push it soon.

rralf commented:
So I just implemented 8a4cd618450c4282cc38893c6dc574e298b5929b and 4ec2d6dd82c585eb1795b72ea3f0828752d0d1b8. You might want to have a look at them. Give me a couple of minutes, I'll rebase your latest push and add my local changes.

rralf commented:
Ok, so I made some API changes that are already integrated on the next branch. Those API changes make it much easier to load a LinuxMaintainers instance given by any revision. You could do, for example:

mtrs = load_maintainers(repo, 'v5.4-rc3')
mtrs = load_maintainers(repo, '1a2b3c4d5d6e')
mtrs = load_maintainers(repo, 'master')
mtrs = load_maintainers(repo, 'HEAD')

They all will work. This makes life a lot easier. I used this chance to implement a get_blob routine in the Repository class. It takes a revision and a filename and returns a blob. Can be useful for many cases.

Given that, have a look at this branch:
https://github.com/lfd/PaStA/tree/basak

It's your tree, rebased to latest next. I did following changes:

- The version is no longer encoded in the filename. It was a silly idea (by me).
- The version is now encoded in the file itself: The version is simply the first line
- --revision can always override the version, even if --filter is given
- If --revision is not set, the script takes the first line of the file
- If --filter is not set, it will default to HEAD
- It will finally log which version it chose

See 0320d549d8f07da8aac278c561aea1b897546f28.

basakerdamar commented:
> Basak, the feature listing
> 
> RELEVANT: list of the most general file patterns such that all files with that pattern are included in the relevant list and are in the entry.
> 
> IRRELEVANT: list of the most general file pattern such that all files under that directory are excluded from the relevant list, but are in the entry.
> 
> is still missing, right?
> 
> If it is not clear at the moment, let us just the dump the whole list of files included and excluded and go through a few examples to see how that can be nicely aggregated.

I added my changes regarding this feature. I removed my old aggregation which lists the common (sub-)directories instead of listing everything, as it was causing confusion when cross-checking relevant and irrelevant lists. Right now, the script plainly lists directories in the filter file list and out of the filter file list.

rralf commented:
Hi Basak,

thanks for the changes! Could you please rebase your branch to the latest version of the next branch of PaStA? We made a lot of changes upstream in the meanwhile.

Thanks!

rralf commented:
Hi Basak,

that was a merge with next. The history is now pretty confusing. Please [rebase](https://git-scm.com/docs/git-rebase) to the latest next branch, so that we get a linear history, iow, w/o criss-cross merges. You can also squash / remove commits that we don't need any longer.

Thanks!

basakerdamar commented:
Hi Ralf, 
Thank you for your help. I re-ordered and squashed my commits so that there is one commit for stubs to pasta and one commit with the script itself, like it was before.

rralf commented:
Perfect, thanks. Makes it much easier for me to review. Will have a look at it tomorrow.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,286 @@
+
+import sys
+import os
+import pygit2
```
nit: please order alphabetically

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,286 @@
+from csv import writer
+from logging import getLogger
+from argparse import ArgumentParser
+from multiprocessing import Pool, cpu_count
```
same here

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,286 @@
+    return [item for sublist in nested_list for item in sublist]
+
+def walk_commit_tree(tree):
+    results = []
```
An object can't appear twice in the list. So
results = set()
would be the appropriate container.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,286 @@
+    return results
+
+def get_tree(repo, revision):
+    if revision != 'HEAD':
```
Please reverse the logic:

```
if revision == 'HEAD':
    return repo.revparse_single(repo.head.name).tree

if revision.startswith...
```

Makes it easier to read.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,286 @@
+            commit_hash = repo[revision].target
+            tree = repo[commit_hash].tree
+        except ValueError:
+            raise
```
If we just raise the exception again, the the try...except block becomes superfluous. Remove it.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,286 @@
+    return sorted(dictionary.items(), key=lambda item:item[1].loc, reverse=True)
+
+def read_file(filename):
+    return _tmp_tree[filename].data.decode('iso8859')
```
single-caller function. please fold it into get_filesize as long as there are no other users.

rralf replied:
And why do you use iso8859? What if a file is encoded in UTF-8? There are quite some.

But if you just need to count for line feeds, there's no need to decode the bytesarray at all.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,286 @@
+def get_filesize(filename):
+    f = read_file(filename)
+    try:
+        return (filename, (f.count('\n'), len(f)))
```
instead of len(f), you should use _tmp_tree[data].size

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,286 @@
+        return (filename, (f.count('\n'), len(f)))
+    # Empty file:
+    except:
+        return (filename, (0, 0))
```
When does an exception occur? This looks odd to me.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,286 @@
+        subsystems = all_maintainers.get_subsystems_by_file(filename)
+        for subsystem in subsystems:
+            maintainers = all_maintainers.get_maintainers(subsystem)
+            for maintainer in flatten(maintainers):
```
Uaah, no! :-)

This is not correct.

all_maintainers.get_maintainers(subsystem)

[get_maintainers](https://github.com/lfd/PaStA/blob/master/pypasta/LinuxMaintainers.py#L157) _always_ gives you a tripe (lists, maintainers, reviewers). I'm still not quite sure what you're trying to achieve, but probably, you only want to access the maintainers.

(I have to admin, maybe get_maintainers should be refactored to get_responsible or something like that)

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,286 @@
+            for maintainer in flatten(maintainers):
+                if not (type(maintainer) == str
+                        or maintainer[0] == ''
+                        or maintainer[1] == ''):
```
Then, this become superfluous.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,286 @@
+            for item in line[1:]:
+                line_s += str(item).ljust(12)
+            if len(line_s) > columns:
+                print(line_s[0:columns-3]+"...")
```
Ok, I don't know what kind of magic happens here, but we first need to address the points above.

rralf commented on bin/pasta_show_maintainers.py:
```diff
@@ -0,0 +1,286 @@
+        self.byte += byte
+        if filter:
+            self.loc_filt += loc
+            self.byte_filt += byte
```
Ok, lots of magic that I don't understand yet, but we need to solve the other issues first.

basakerdamar replied:
I meant this to be a counter for lines, keeping track of in-list and and total separately. Though I do not need the file name anymore, I fixed this in my last push.

rralf reviewed (changes requested):
Hi Basak,

thanks for the new version!

Please find my review attached. I already addressed some of the comments in order to get the script running. Please find those patches on the [basak](https://github.com/lfd/PaStA/tree/basak)-branch. You can integrate the topmost patch to your branch.

Most important thing: I guess you misunderstood the result of get_maintainers(). It returns you a triple of lists, maintainers, reviewers. You try to somehow (erroneously) flatten(?) the result.

Besides that, I tried to run `pasta show_maintainers --filter test` with the following content of test:
```
v4.7-rc3
MAINTAINERS
scripts/get_maintainers.pl
Makefile
```

As result, I got:
```2020-06-02 22:54:51,096 how_maintainers INFO     Working on kernel revision v4.7-rc3
Maintainers                                        LoC         LoC in list In list(%)  Entries of maintainer
linus torvalds  <torvalds@linux-foundation.org>    21716623    14399       0.07        THE REST    
michal marek  <mmarek@suse.com>                    13786       1667        12.09       COCCINELLE/Semantic Patches (SmPL), KERNEL BUILD + files below scripts...
2020-06-02 22:56:43,880 PaStA           INFO     Shutting down


```

... which somehow looks odd to me. Is the output intended?


One last thing: I'm somehow struggling with the name `show_maintainers`. I bet we can find a better name for the script. What we actually do is printing statistics of responsibilities of either the whole kernel, or a certain subset of the kernel. Maybe a better name would be `subsystem_stats` or maybe `maintainers_stats`. I happily accept better suggestions! :-)

Thanks!
  Ralf

rralf commented:
https://github.com/lfd/PaStA/commit/cee96b78127eb950e09768107f890c3b85b3670c

So this is the commit with my fixups. Mainly, I improved walk_commit_tree (use sets instead of lists), simplified get_tree() and corrected get_file_size() (we don't need decoding of strings, and we mustn't use len() to get the size of a file).

basakerdamar commented:
> Hi Basak,
> 
> thanks for the new version!
> 
> Please find my review attached. I already addressed some of the comments in order to get the script running. Please find those patches on the [basak](https://github.com/lfd/PaStA/tree/basak)-branch. You can integrate the topmost patch to your branch.
> 
> Most important thing: I guess you misunderstood the result of get_maintainers(). It returns you a triple of lists, maintainers, reviewers. You try to somehow (erroneously) flatten(?) the result.
> 
> Besides that, I tried to run `pasta show_maintainers --filter test` with the following content of test:
> 
> ```
> v4.7-rc3
> MAINTAINERS
> scripts/get_maintainers.pl
> Makefile
> ```
> 
> As result, I got:
> 
> ```
> Maintainers                                        LoC         LoC in list In list(%)  Entries of maintainer
> linus torvalds  <torvalds@linux-foundation.org>    21716623    14399       0.07        THE REST    
> michal marek  <mmarek@suse.com>                    13786       1667        12.09       COCCINELLE/Semantic Patches (SmPL), KERNEL BUILD + files below scripts...
> 2020-06-02 22:56:43,880 PaStA           INFO     Shutting down
> ```
> 
> ... which somehow looks odd to me. Is the output intended?
> 
> One last thing: I'm somehow struggling with the name `show_maintainers`. I bet we can find a better name for the script. What we actually do is printing statistics of responsibilities of either the whole kernel, or a certain subset of the kernel. Maybe a better name would be `subsystem_stats` or maybe `maintainers_stats`. I happily accept better suggestions! :-)
> 
> Thanks!
> Ralf

Hi Ralf,

Thank you for the changes!

The output was what I intended, although now 'THE REST' is excluded in each view. Hopefully this way it will be more meaningful.

I also added the --files option to see the entries for each file and changed the script's name to maintainers_stats.

rralf commented:
Then you need to explain the output to me. There are surely more maintainers than just Michal Marek that are responsible for the three files.

rralf commented:
Okay, I still don't get it. What's the semantics of `maintainers_state --filter`? What does it tell me? With your new version, it prints now nothing (same example as above):
```
2020-06-04 12:48:44,590 PaStA           INFO     Cmdline: ./pasta maintainers_stats --filter ./test
2020-06-04 12:48:44,590 pypasta.Config  INFO     Active configuration: linux
2020-06-04 12:48:44,635 intainers_stats INFO     Working on kernel revision v4.7-rc3
Maintainers    LoC         LoC in list In list(%)  Entries of maintainer
2020-06-04 12:51:02,258 PaStA           INFO     Shutting down
```

basakerdamar commented:
> Then you need to explain the output to me. There are surely more maintainers than just Michal Marek that are responsible for the three files.

Sorry, I thought the confusion was about the format/fields of the output. There are clearly some lines missing in this result. I am trying to find the problem causing the missing maintainers, I pushed another fix too.

rralf commented on bin/pasta_maintainers_stats.py:
```diff
@@ -0,0 +1,303 @@
+                all_maintainers, file_sizes, filter_by_files)
+        else:
+            results, _, _= get_counts_by_entry(all_maintainers, file_sizes)
+    elif args.files:
```
`--entries` and `--files` are mutual exclusive, right?

So if I call `maintainers-stats --entries --files`, then --files will by ignored.

In this case, it's probably better to make it clear by introducing a mode. E.g.
```
maintainers_stats --mode entries
maintainers_stats --mode files
```

We even have three modes, right? entries, files, and maintainers, if i see that correctly. We should make that clear in the commandline arguments.

And it's probably better to rename entries to subsystems.

bulwahn replied:
Yes, we should introduce some option like that, I would call it --group-by or --aggregate-on ... entries (or section, not subsystem)/files/maintainers instead of --mode.

rralf replied:
Yep, sounds good. And use, e.g., sections (formerly knows as entries) as default value.

rralf commented on bin/pasta_maintainers_stats.py:
```diff
@@ -0,0 +1,303 @@
+
+def get_entries_by_files(all_maintainers, filenames):
+    filenames = sorted(filenames, key=lambda item:os.path.basename(item), 
+                        reverse=True)
```
Could you explain what this routine does? I don't unterstand the option --files.

bulwahn replied:
Very quickly, it gives you a list of file to entry mapping.

Here is quick dump of the csv output, dumped into Google Sheets and formated and reordered (alphabetically): https://docs.google.com/spreadsheets/d/13O5bstZpENC04EvrTziYKe-DilyLwTH6RQ7sWukS82c/edit?usp=sharing

You can quickly see there is structure and substructure of responsibilities according to the directory etc., maybe we then also find way to visualize that nicely...

rralf replied:
Thanks, got it. Then I'd like to ask for the following change:

AFAICT, the ordering of the filenames does not matter for the evaluation, right? You alphabetically order filenames for reasons of visualisation.

Could you please remove the ordering from this routine? Ordering is for visualisation and not for processing. So we should, for example, return a dictionary {filename: ([entries], loc, size)}. The caller can then do further postprocessing, like ordering things.
The reason: I was wondering why the order plays a role while iterating over the filenames. If we postpone the ordering, it gets clear that it's just about visualisation.

rralf reviewed (changes requested):
Hi Basak,

cool, things are getting much clearer now! Again, have a look at the upstream [basak branch](https://github.com/lfd/PaStA/tree/basak). I added two commits that simplify things.

The first commit simplifies the logic inside the evaluators (we can drop a lot of duplicated code to simplify things), and the second commit simplifies the structure of the caller-site.

However, we aren't yet there. What I realised during the refactoring is that get_counts_by_entry and get_counts_by_maintainer basically almost have the same structure. We should consolidate those routines to one single routine, and add an additional argument that selects wether the user wants to consider maintainer or entries.

BTW: With entries, you refer to entries in MAINTAINERS, aka. subsystems, right? If so, we should use the term subsystem instead of entry.

I also made some changes to the Counter-class. Please check if I missed something, I'm still not totally aware of its internal logic.

Thanks!
  Ralf

bulwahn commented:
@basakerdamar, here is some feedback to --files option on fdcbc61:
  - order files alphabetically by default
  - how are the entries ordered? Can we order them by size by default?
  - let us drop LoC, file size by default and create an option for adding those columns if requested.

bulwahn commented:
> BTW: With entries, you refer to entries in MAINTAINERS, aka. subsystems, right? If so, we should use the term subsystem instead of entry.
> 
Yes, with entries we refer to entries in MAINTAINERS and I disagree with the suggestion to use the term subsystem. Not every entry in MAINTAINERS is a subsystem (in kernel developer lingo). Look at the wording in MAINTAINERS, we can use "section" (as in "Descriptions of section entries" in MAINTAINERS, https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/MAINTAINERS#n80) or "entries" as in  "When adding to this list, please keep the entries in alphabetical order.", https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/MAINTAINERS#n142). Hence, I early decided for entries in my wording and kernel developers did not disagree on my patches on MAINTAINERS. So, either section or entry, as far as I see it. Entry is not to bad for this tool.

rralf commented:
Ok, but then `get_maintainers.pl --subsystem` should then consistently be changed to `get_maintainers.pl --entries` instead.

I get your point, but from my understanding, and in terms of get_maintainers.pl, the main "user" of MAINTAINERS, a section in MAINTAINERS is a subsystem. For me, the term 'entry' is a bit misleading, as I rather associate a single line with it instead of a whole section.

In the document that you pointed to, they mention the term 'section entry'. I'd rather use this term, or the short variant `section` then.

bulwahn commented:
> Ok, but then `get_maintainers.pl --subsystem` should then consistently be changed to `get_maintainers.pl --entries` instead.
>
Consistency, yes!
 
> I get your point, but from my understanding, and in terms of get_maintainers.pl, the main "user" of MAINTAINERS, a section in MAINTAINERS is a subsystem. For me, the term 'entry' is a bit misleading, as I rather associate a single line with it instead of a whole section.
>

But simply go out and say you work on the CLANG-FORMAT FILE subsystem, people are going to say that is not a subsystem, although it deserve to be a section in MAINTAINERS.
  
> In the document that you pointed to, they mention the term 'section entry'. I'd rather use this term, or the short variant `section` then.

Okay, I am happy to settle for section. Let us get the whole wording right and consistent for the commandline and the whole code with using the word section.

basakerdamar commented:
> Then you need to explain the output to me. There are surely more maintainers than just Michal Marek that are responsible for the three files.

Hi Ralf,
Short question for debugging:
For your filter example,
from 
```
LinuxMaintainers.get_subsystems_by_file('Makefile')
```
I get:
```
{'THE REST', 'KERNEL BUILD + files below scripts/ (unless maintained elsewhere)'}
```
and
```
LinuxMaintainers.get_maintainers('KERNEL BUILD + files below scripts/ (unless maintained elsewhere)')
```
gives
```
({'linux-kbuild@vger.kernel.org'}, [('michal marek ', 'mmarek@suse.com')], [])
```
Same for scripts/get_maintainer.pl, I only get one maintainer:
```
{'GET_MAINTAINER SCRIPT', 'THE REST'}
```
and for 'GET_MAINTAINER SCRIPT', 
```
(set(), [('joe perches ', 'joe@perches.com')], [])
```
The results is therefore only two maintainers. Could it be that LinuxMaintainers class gives me too few maintainers or am I using the class in a wrong way?

rralf commented on bin/pasta_maintainers_stats.py:
```diff
@@ -0,0 +1,275 @@
+                        'enter the file name for the file containing the list '
+                        'of files to filter by.')
+    parser.add_argument('--group-by', type=str, default='sections',
+                        help='files option '
```
Add the choices={'entries', 'section', 'files'} argument. argparse will do then sanity checks for arguments.

rralf commented on bin/pasta_maintainers_stats.py:
```diff
@@ -0,0 +1,275 @@
+        results = get_sections_by_files(all_maintainers, victims)
+    else:
+        parser.print_help()
+        sys.exit()
```
We can leave out the whole else case if we feed the argparser with the choices argument

rralf commented on bin/pasta_maintainers_stats.py:
```diff
@@ -0,0 +1,275 @@
+                        'maintainers option groups files by sections, displays '
+                        'each maintainer ordered by lines of code they are '
+                        'responsible for\n'
+                        'Default is sections if no value is specified')
```
Please separate sentences with interpunctation. And have a look at the output that pasta maintainers_stats -h generates, this help section is really hard to read.

rralf commented on bin/pasta_maintainers_stats.py:
```diff
@@ -0,0 +1,275 @@
+    elif args.group_by == 'maintainers':
+        title = ['Maintainers', 'LoC']
+        results, maintainer_to_section = \
+         get_counts(all_maintainers, file_sizes, filter_by_files, args.group_by)
```
```
title = [args.group_by, 'LoC']
if args.group_by in {'sections', 'maintainers'}:
    result = get_counts(all_maintainers, file_sizes, filter_by_files, args.group_by)
```

And we get rid of the remaining redundancy.

rralf commented on bin/pasta_maintainers_stats.py:
```diff
@@ -0,0 +1,275 @@
+    if group_by == 'sections':
+        return decreasing_order(counts), filename_excluded, filenames_filt
+    else:
+        return decreasing_order(counts), maintainer_to_section
```
Mhhm. No... Returning different signatures shifts the problem to the caller-site. I wonder if we can somehow get a common datatype for both cases...

rralf commented on bin/pasta_maintainers_stats.py:
```diff
@@ -0,0 +1,275 @@
+        # all_maintainers.subsystems[section].files
+        sections = sorted(sections, 
+         key=lambda section: len(all_maintainers.subsystems[section].files))
+        results.append([filename, sections])
```
I don't get this ordering. What do you try to achieve? Sort subsystems by their size?

rralf reviewed (commented):
CC: Pia Eichinger @ditzlike 

Basak, Pia is currently working on improving LinuxMaintainers.py. Pia, could you please check that? I'll also have a look at it later.

rralf commented:
Basak, thanks for the PR. Will continue working on it and get back to you.

rralf committed:
ralf rewrite logic

rralf commented:
@basakerdamar For your questions regarding the Maintainers...

Let's take v5.7-rc3 as reference (randomly chosen).

```
./scripts/get_maintainers.pl --subsystem -f ./Makefile
Masahiro Yamada <masahiroy@kernel.org> (maintainer:KERNEL BUILD + files below scripts/ (unless mai...)
Michal Marek <michal.lkml@markovi.net> (maintainer:KERNEL BUILD + files below scripts/ (unless mai...)
linux-kbuild@vger.kernel.org (open list:KERNEL BUILD + files below scripts/ (unless mai...)
linux-kernel@vger.kernel.org (open list)
clang-built-linux@googlegroups.com (open list:CLANG/LLVM BUILD SUPPORT)
KERNEL BUILD + files below scripts/ (unless maintained elsewhere)
THE REST
CLANG/LLVM BUILD SUPPORT
```

If I run in PaStA:

```
    maintainers_version = load_maintainers(config, {'v5.7-rc3'})
    maintainers = maintainers_version['v5.7-rc3']

    sections = maintainers.get_subsystems_by_file('Makefile')
    for section in sections:
        mtrs = maintainers.get_maintainers(section)
        print(mtrs)
```
It reports:
```
({'linux-kernel@vger.kernel.org'}, [('linus torvalds', 'torvalds@linux-foundation.org')], [])
({'linux-kbuild@vger.kernel.org'}, [('masahiro yamada', 'masahiroy@kernel.org'), ('michal marek', 'michal.lkml@markovi.net')], [])
```

which has two slight differences:
  - get_maintainers.pl will not show Linus Torvalds (unless --git-chief-penguins (yes, really) is specified)
  - PaStA misses the CLANG/LLVM BUILD SUPPORT section. Reason: LinuxMaintainers.py currently does not evaluate the K:-fields in MAINTAINERS.

So from my side, I would say everything looks fine.

Does that help?

rralf commented:
@ditzlike Note for us: PaStA lacks understanding for the K:-field.

ditzlike commented:
Noted. I'll see if I get to fix it this weekend

rralf commented:
Happy hacking. But this is going to be really challenging. For the moment we should probably skip the K-tags, and see how good our results are without them.

basakerdamar committed:
Additional fields, left aligned formatting
